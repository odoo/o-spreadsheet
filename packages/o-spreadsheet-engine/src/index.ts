export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
