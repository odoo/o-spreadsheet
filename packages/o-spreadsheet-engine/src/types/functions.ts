import { Arg, CellPosition, CellValue, FunctionResultObject, Matrix, UID } from "./index";
import { Locale } from "./locale";

export type ArgType =
  | "ANY"
  | "BOOLEAN"
  | "NUMBER"
  | "STRING"
  | "DATE"
  | "RANGE"
  | "RANGE<BOOLEAN>"
  | "RANGE<NUMBER>"
  | "RANGE<DATE>"
  | "RANGE<STRING>"
  | "RANGE<ANY>"
  | "META"
  | "RANGE<META>";

export interface ArgDefinition {
  acceptMatrix?: boolean;
  acceptMatrixOnly?: boolean;
  acceptErrors?: boolean;
  repeating?: boolean;
  optional?: boolean;
  description: string;
  name: string;
  type: ArgType[];
  default?: boolean;
  defaultValue?: any;
  proposalValues?: ArgProposal[];
}

export type ArgProposal = { value: CellValue; label?: string };
export type ComputeFunction<R> = (this: EvalContext, ...args: Arg[]) => R;

export interface Functions {
  compute: ComputeFunction<
    FunctionResultObject | Matrix<FunctionResultObject> | CellValue | Matrix<CellValue>
  >;
  description: string;
  category?: string;
  args: ArgDefinition[];
  isExported?: boolean;
  hidden?: boolean;
}

export type FunctionDescription = Functions & {
  name: string;
  minArgRequired: number;
  maxArgPossible: number;
  nbrArgRepeating: number;
  nbrArgOptional: number;
};
export type EvalContext = {
  __originSheetId: UID;
  __originCellPosition?: CellPosition;
  locale: Locale;
  getters: Getters;
  [key: string]: any;
  updateDependencies?: (position: CellPosition) => void;
  addDependencies?: (position: CellPosition, ranges: Range[]) => void;
  debug?: boolean;
  lookupCaches?: LookupCaches;
};
/**
 * used to cache lookup values for linear search
 **/
export type LookupCaches = {
  forwardSearch: Map<unknown, Map<CellValue, number>>;
  reverseSearch: Map<unknown, Map<CellValue, number>>;
};
