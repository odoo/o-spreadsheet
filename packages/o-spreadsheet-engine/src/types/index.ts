export * from "./base";
export * from "./errors";
export * from "./functions";
export * from "./history";
export * from "./isMatrix";
export * from "./locale";
export * from "./validator";
