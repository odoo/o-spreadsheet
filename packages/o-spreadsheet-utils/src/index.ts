export * from "./translation";
export * from "./uuid";
