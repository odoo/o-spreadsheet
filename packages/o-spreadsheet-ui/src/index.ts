import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}