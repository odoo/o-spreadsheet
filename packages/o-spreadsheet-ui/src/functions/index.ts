import { CellComposerStore } from "../components/composer/composer/cell_composer_store";
import { tokenColors } from "../constants";
import { EnrichedToken } from "../formulas/composer_tokenizer";
import {
  AutoCompleteProposal,
  autoCompleteProviders,
} from "../registries/auto_completes/auto_complete_registry";
import { Registry } from "../registries/registry";
import { _t } from "../translation";
import {
  AddFunctionDescription,
  Arg,
  ArgDefinition,
  CellValue,
  ComputeFunction,
  EvalContext,
  FunctionDescription,
  FunctionResultObject,
  Matrix,
  isMatrix,
} from "../types";
import { BadExpressionError, EvaluationError } from "../types/errors";
import { addMetaInfoFromArg, argTargeting, validateArguments } from "./arguments";
import { applyVectorization, isEvaluationError, matrixForEach, matrixMap } from "./helpers";
import * as array from "./module_array";
import * as misc from "./module_custom";
import * as database from "./module_database";
import * as date from "./module_date";
import * as engineering from "./module_engineering";
import * as filter from "./module_filter";
import * as financial from "./module_financial";
import * as info from "./module_info";
import * as logical from "./module_logical";
import * as lookup from "./module_lookup";
import * as math from "./module_math";
import * as operators from "./module_operators";
import * as parser from "./module_parser";
import * as statistical from "./module_statistical";
import * as text from "./module_text";
import * as web from "./module_web";

export { arg } from "./arguments";

type Functions = { [functionName: string]: AddFunctionDescription };
type Category = { name: string; functions: Functions };
const categories: Category[] = [
  { name: _t("Array"), functions: array },
  { name: _t("Database"), functions: database },
  { name: _t("Date"), functions: date },
  { name: _t("Filter"), functions: filter },
  { name: _t("Financial"), functions: financial },
  { name: _t("Info"), functions: info },
  { name: _t("Lookup"), functions: lookup },
  { name: _t("Logical"), functions: logical },
  { name: _t("Math"), functions: math },
  { name: _t("Misc"), functions: misc },
  { name: _t("Operator"), functions: operators },
  { name: _t("Statistical"), functions: statistical },
  { name: _t("Text"), functions: text },
  { name: _t("Engineering"), functions: engineering },
  { name: _t("Web"), functions: web },
  { name: _t("Parser"), functions: parser },
];

const functionNameRegex = /^[A-Z0-9\_\.]+$/;

//------------------------------------------------------------------------------
// Function registry
//------------------------------------------------------------------------------

export class FunctionRegistry extends Registry<FunctionDescription> {
  mapping: {
    [key: string]: ComputeFunction<Matrix<FunctionResultObject> | FunctionResultObject>;
  } = {};

  add(name: string, addDescr: AddFunctionDescription) {
    name = name.toUpperCase();
    if (name in this.content) {
      throw new Error(`${name} is already present in this registry!`);
    }
    return this.replace(name, addDescr);
  }

  replace(name: string, addDescr: AddFunctionDescription) {
    name = name.toUpperCase();
    if (!functionNameRegex.test(name)) {
      throw new Error(
        _t(
          "Invalid function name %s. Function names can exclusively contain alphanumerical values separated by dots (.) or underscore (_)",
          name
        )
      );
    }
    const descr = addMetaInfoFromArg(name, addDescr);
    validateArguments(descr);
    this.mapping[name] = createComputeFunction(descr);
    super.replace(name, descr);
    return this;
  }
}

export const functionRegistry: FunctionRegistry = new FunctionRegistry();

for (const category of categories) {
  const fns = category.functions;
  for (let name in fns) {
    const addDescr = fns[name];
    addDescr.category = addDescr.category || category.name;
    name = name.replace(/_/g, ".");
    functionRegistry.add(name, { isExported: false, ...addDescr });

    createAutocompleteArgumentsProvider(name, addDescr.args);
  }
}

//------------------------------------------------------------------------------
// CREATE COMPUTE FUNCTION
//------------------------------------------------------------------------------

function createComputeFunction(
  descr: FunctionDescription
): ComputeFunction<Matrix<FunctionResultObject> | FunctionResultObject> {
  function vectorizedCompute(
    this: EvalContext,
    ...args: Arg[]
  ): FunctionResultObject | Matrix<FunctionResultObject> {
    const acceptToVectorize: boolean[] = [];

    const getArgToFocus = argTargeting(descr, args.length);
    //#region Compute vectorisation limits
    for (let i = 0; i < args.length; i++) {
      const argIndex = getArgToFocus(i) ?? -1;
      const argDefinition = descr.args[argIndex];
      const arg = args[i];
      if (!isMatrix(arg) && argDefinition.acceptMatrixOnly) {
        throw new BadExpressionError(
          _t(
            "Function %s expects the parameter '%s' to be reference to a cell or range.",
            descr.name,
            (i + 1).toString()
          )
        );
      }
      acceptToVectorize.push(!argDefinition.acceptMatrix);
    }

    return applyVectorization(errorHandlingCompute.bind(this), args, acceptToVectorize);
  }

  function errorHandlingCompute(
    this: EvalContext,
    ...args: Arg[]
  ): Matrix<FunctionResultObject> | FunctionResultObject {
    for (let i = 0; i < args.length; i++) {
      const arg = args[i];
      const getArgToFocus = argTargeting(descr, args.length);
      const argDefinition = descr.args[getArgToFocus(i) || i];

      // Early exit if the argument is an error and the function does not accept errors
      // We only check scalar arguments, not matrix arguments for performance reasons.
      // Casting helpers are responsible for handling errors in matrix arguments.
      if (!argDefinition.acceptErrors && !isMatrix(arg) && isEvaluationError(arg?.value)) {
        return arg;
      }
    }
    try {
      return computeFunctionToObject.apply(this, args);
    } catch (e) {
      return handleError(e, descr.name);
    }
  }

  function computeFunctionToObject(
    this: EvalContext,
    ...args: Arg[]
  ): FunctionResultObject | Matrix<FunctionResultObject> {
    if (this.debug) {
      // eslint-disable-next-line no-debugger
      debugger;
    }
    const result = descr.compute.apply(this, args);

    if (!isMatrix(result)) {
      if (typeof result === "object" && result !== null && "value" in result) {
        replaceFunctionNamePlaceholder(result, descr.name);
        return result;
      }
      return { value: result };
    }

    if (typeof result[0][0] === "object" && result[0][0] !== null && "value" in result[0][0]) {
      matrixForEach(result as Matrix<FunctionResultObject>, (result) =>
        replaceFunctionNamePlaceholder(result, descr.name)
      );
      return result as Matrix<FunctionResultObject>;
    }

    return matrixMap(result as Matrix<CellValue>, (row) => ({ value: row }));
  }

  return vectorizedCompute;
}

export function handleError(e: unknown, functionName: string): FunctionResultObject {
  // the error could be an user error (instance of EvaluationError)
  // or a javascript error (instance of Error)
  // we don't want block the user with an implementation error
  // so we fallback to a generic error
  if (hasStringValue(e) && isEvaluationError(e.value)) {
    if (hasStringMessage(e)) {
      replaceFunctionNamePlaceholder(e, functionName);
    }
    return e;
  }
  console.error(e);
  return new EvaluationError(
    implementationErrorMessage + (hasStringMessage(e) ? " " + e.message : "")
  );
}

function hasStringValue(obj: unknown): obj is { value: string } {
  return (
    (obj as { value: string })?.value !== undefined &&
    typeof (obj as { value: string }).value === "string"
  );
}

function replaceFunctionNamePlaceholder(
  functionResult: FunctionResultObject,
  functionName: string
) {
  // for performance reasons: change in place and only if needed
  if (functionResult.message?.includes("[[FUNCTION_NAME]]")) {
    functionResult.message = functionResult.message.replace("[[FUNCTION_NAME]]", functionName);
  }
}

export const implementationErrorMessage = _t(
  "An unexpected error occurred. Submit a support ticket at odoo.com/help."
);

function hasStringMessage(obj: unknown): obj is { message: string } {
  return (
    (obj as { message: string })?.message !== undefined &&
    typeof (obj as { message: string }).message === "string"
  );
}

function createAutocompleteArgumentsProvider(formulaName: string, args: ArgDefinition[]) {
  for (let i = 0; i < args.length; i++) {
    const proposalValues = args[i].proposalValues;
    if (proposalValues === undefined || proposalValues.length === 0) {
      continue;
    }

    const getProposals = (tokenAtCursor: EnrichedToken) => {
      const functionContext = tokenAtCursor.functionContext;
      if (
        !functionContext ||
        functionContext.parent.toUpperCase() !== formulaName.toUpperCase() ||
        functionContext.argPosition !== i
      ) {
        return;
      }

      const proposals: AutoCompleteProposal[] = [];
      let text = "";
      for (const { value, label } of proposalValues) {
        switch (typeof value) {
          case "string":
            text = `"${value}"`;
            break;
          case "number":
            text = value.toString();
            break;
          case "boolean":
            text = value ? "TRUE" : "FALSE";
            break;
          default:
        }

        proposals.push({
          text,
          description: label,
          htmlContent: [
            {
              value: text,
              color: typeof value === "string" ? tokenColors.STRING : tokenColors.NUMBER,
            },
          ],
          fuzzySearchKey: text,
          alwaysExpanded: true,
        });
      }

      return proposals;
    };

    autoCompleteProviders.add(`${formulaName}_function_${args[i].name}_argument_proposals`, {
      sequence: 50,
      autoSelectFirstProposal: true,
      selectProposal: insertTokenAtArgStartingPosition,
      getProposals,
    });
  }
}

/**
 * Perform the autocomplete of the composer by inserting the value
 * at the cursor position, replacing the current token if necessary.
 * Must be bound to the autocomplete provider.
 */
export function insertTokenAtArgStartingPosition(
  this: { composer: CellComposerStore },
  tokenAtCursor: EnrichedToken,
  value: string
) {
  let start = tokenAtCursor.end;
  const end = tokenAtCursor.end;
  if (!["LEFT_PAREN", "ARG_SEPARATOR"].includes(tokenAtCursor.type)) {
    // replace the whole token
    start = tokenAtCursor.start;
  }
  this.composer.stopComposerRangeSelection();
  this.composer.changeComposerCursorSelection(start, end);
  this.composer.replaceComposerCursorSelection(value);
}
