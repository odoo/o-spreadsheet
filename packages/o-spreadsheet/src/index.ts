import * as UI from "@odoo/o-spreadsheet-ui";


export { UI };