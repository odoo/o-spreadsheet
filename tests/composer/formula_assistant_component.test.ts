import { setTranslationMethod } from "../../src";
import { ComposerStore } from "../../src/components/composer/composer/composer_store";
import { arg, functionRegistry } from "../../src/functions/index";
import { Store } from "../../src/store_engine";
import { _t } from "../../src/translation";
import { DEFAULT_LOCALE } from "../../src/types";
import { registerCleanup } from "../setup/jest.setup";
import { updateLocale } from "../test_helpers/commands_helpers";
import { keyDown, keyUp } from "../test_helpers/dom_helper";
import {
  ComposerWrapper,
  clearFunctions,
  mountComposerWrapper,
  nextTick,
  restoreDefaultFunctions,
  typeInComposerHelper,
} from "../test_helpers/helpers";
jest.mock("../../src/components/composer/content_editable_helper.ts", () =>
  require("../__mocks__/content_editable_helper")
);

let composerEl: Element;
let fixture: HTMLElement;
let parent: ComposerWrapper;
let composerStore: Store<ComposerStore>;

async function typeInComposer(text: string, fromScratch: boolean = true) {
  if (fromScratch) {
    parent.startComposition();
  }
  await typeInComposerHelper("div.o-composer", text, false);
}

beforeEach(async () => {
  ({ fixture, parent } = await mountComposerWrapper());
  // start composition
  parent.startComposition();
  await nextTick();
  composerEl = fixture.querySelector("div.o-composer")!;
  composerStore = parent.env.getStore(ComposerStore);
});

describe("formula assistant", () => {
  beforeAll(() => {
    clearFunctions();
    functionRegistry.add("FUNC0", {
      description: "func without args",
      args: [],
      compute: () => 1,
    });
    setTranslationMethod(
      (str, ...values) => str,
      () => false
    );
    functionRegistry.add("FUNC1", {
      description: "func1 def",
      args: [arg("f1Arg1 (any)", "f1 Arg1 def"), arg("f1Arg2 (any)", _t("f1 Arg2 def"))],
      compute: () => 1,
    });
    setTranslationMethod((str, ...values) => str);
    functionRegistry.add("FUNC2", {
      description: "func2 def",
      args: [
        arg("f2Arg1 (any)", "f2 Arg1 def"),
        arg("f2Arg2 (any, optional, default=TRUE)", "f2 Arg2 def"),
      ],
      compute: () => 1,
    });
    functionRegistry.add("FUNC3", {
      description: "func3 def",
      args: [
        arg("f3Arg1 (any)", "f3 Arg1 def"),
        arg("f3Arg2 (any, optional, repeating)", "f3 Arg2 def"),
      ],
      compute: () => 1,
    });
    functionRegistry.add("UPTOWNFUNC", {
      description: "a Bruno Mars song ?",
      args: [
        arg("f4Arg1 (any)", "f4 Arg1 def"),
        arg("f4Arg2 (any, optional, repeating)", "f4 Arg2 def"),
        arg("f4Arg3 (any, optional, repeating)", "f4 Arg3 def"),
      ],
      compute: () => 1,
    });
  });

  afterAll(() => {
    restoreDefaultFunctions();
  });

  describe("appearance", () => {
    test("empty not show autocomplete", async () => {
      await typeInComposer("");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("= do not show formula assistant", async () => {
      await typeInComposer("=");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("=FUNC1( show formula assistant", async () => {
      await typeInComposer("=FUNC1(");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(1);
    });

    test("=func1( show formula assistant", async () => {
      await typeInComposer("=func1(");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(1);
    });

    test("FUNC1( do not show formula assistant", async () => {
      await typeInComposer("FUNC1");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("=FUNC1 do not show formula assistant", async () => {
      await typeInComposer("=FUNC1");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("=FUN( do not show formula assistant (nothing matches FUN)", async () => {
      await typeInComposer("=FUN(");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("=FUNC1) do not show formula assistant", async () => {
      await typeInComposer("=FUNC1)");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("=FUNC1() do not show formula assistant", async () => {
      await typeInComposer("=FUNC1()");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("=FUNC1(( do not show formula assistant", async () => {
      await typeInComposer("=FUNC1((");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("=FUNC1)( do not show formula assistant", async () => {
      await typeInComposer("=FUNC1)(");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("=FUNC1(() show formula assistant", async () => {
      await typeInComposer("=FUNC1(()");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(1);
    });

    test("=FUNC1()( do not show formula assistant", async () => {
      await typeInComposer("=FUNC1()(");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("=FUNC1(FUNC2( show formula assistant for 2nd function", async () => {
      await typeInComposer("=FUNC1(FUNC2(");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(1);
      expect(fixture.querySelectorAll(".o-formula-assistant-head span")[0].textContent).toBe(
        "FUNC2"
      );
    });

    test("=FUNC1(FUNC2() show formula assistant for 1st function", async () => {
      await typeInComposer("=FUNC1(FUNC2()");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(1);
      expect(fixture.querySelectorAll(".o-formula-assistant-head span")[0].textContent).toBe(
        "FUNC1"
      );
    });

    test("=FUNC1(FUNC2 do not show formula assistant", async () => {
      await typeInComposer("=FUNC1(FUNC2");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("=FUNC1(A2 show formula assistant (A2 is a ref)", async () => {
      await typeInComposer("=FUNC1(A2");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(1);
    });

    test("=FUNC1('a, do not show formula assistant (A2 is a ref)", async () => {
      await typeInComposer("=FUNC1('a,");
      expect(document.activeElement).toBe(composerEl);
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("simple snapshot with =FUNC1(", async () => {
      await typeInComposer("=FUNC1(");
      expect(fixture.querySelector(".o-formula-assistant-container")).toMatchSnapshot();
    });

    test("use arrowKey when selection in a function should not display formula assistant", async () => {
      await typeInComposer("=FUNC1(1,");
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(1);
      await keyDown({ key: "ArrowRight" });
      await keyUp({ key: "ArrowRight" });
      expect(composerStore.currentContent).toBe("=FUNC1(1,B1");
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(0);
    });

    test("use arrowKey during 'editing' mode in a function should display formula assistant", async () => {
      await typeInComposer("=FUNC1(1");
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(1);
      expect(composerStore.editionMode).toBe("editing");
    });

    describe("function definition", () => {
      test("function without argument", async () => {
        await typeInComposer("=FUNC0(");
        expect(fixture.querySelectorAll(".o-formula-assistant-head")[0].textContent).toBe(
          "FUNC0 (  ) "
        );
      });

      test("normal function", async () => {
        await typeInComposer("=FUNC1(");
        expect(fixture.querySelectorAll(".o-formula-assistant-head")[0].textContent).toBe(
          "FUNC1 ( f1Arg1, f1Arg2 ) "
        );
      });

      test("function with default argument", async () => {
        await typeInComposer("=FUNC2(");
        expect(fixture.querySelectorAll(".o-formula-assistant-head")[0].textContent).toBe(
          "FUNC2 ( f2Arg1, [f2Arg2] ) "
        );
      });

      test("function with repeatable argument", async () => {
        await typeInComposer("=FUNC3(");
        expect(fixture.querySelectorAll(".o-formula-assistant-head")[0].textContent).toBe(
          "FUNC3 ( f3Arg1, [f3Arg2, ...] ) "
        );
      });

      test("arguments separator is localized", async () => {
        updateLocale(parent.env.model, { ...DEFAULT_LOCALE, formulaArgSeparator: ";" });
        await typeInComposer("=FUNC1(");
        expect(fixture.querySelectorAll(".o-formula-assistant-head")[0].textContent).toBe(
          "FUNC1 ( f1Arg1; f1Arg2 ) "
        );
      });
    });

    describe("arguments description", () => {
      test("function without argument", async () => {
        await typeInComposer("=FUNC0(");
        expect(fixture.querySelectorAll(".o-formula-assistant-arg")).toHaveLength(0);
      });

      test("normal argument", async () => {
        await typeInComposer("=FUNC1(");
        expect(fixture.querySelectorAll(".o-formula-assistant-arg")).toHaveLength(2);
        expect(fixture.querySelectorAll(".o-formula-assistant-arg div")[2].textContent).toBe(
          "f1Arg2"
        );
        expect(fixture.querySelectorAll(".o-formula-assistant-arg div")[3].textContent).toBe(
          "f1 Arg2 def"
        );
      });

      test("argument description is translated", async () => {
        // set the translation method after creating the argument.
        // This is what actually happens because translations are
        // loaded after the function definition
        setTranslationMethod((str, ...values) => {
          if (str === "f1 Arg2 def") {
            return "translated description";
          }
          return str;
        });
        await typeInComposer("=FUNC1(");
        registerCleanup(() => setTranslationMethod((str, ...values) => str));
        expect(fixture.querySelectorAll(".o-formula-assistant-arg div")[3].textContent).toBe(
          "translated description"
        );
      });

      test("function with default argument", async () => {
        await typeInComposer("=FUNC2(");
        expect(fixture.querySelectorAll(".o-formula-assistant-arg")).toHaveLength(2);
        expect(fixture.querySelectorAll(".o-formula-assistant-arg div")[2].textContent).toBe(
          "f2Arg2 - [optional] default: TRUE"
        );
        expect(fixture.querySelectorAll(".o-formula-assistant-arg div")[3].textContent).toBe(
          "f2 Arg2 def"
        );
      });

      test("function with repeatable argument", async () => {
        await typeInComposer("=FUNC3(");
        expect(fixture.querySelectorAll(".o-formula-assistant-arg")).toHaveLength(2);
        expect(fixture.querySelectorAll(".o-formula-assistant-arg div")[2].textContent).toBe(
          "f3Arg2 - [optional] repeatable"
        );
        expect(fixture.querySelectorAll(".o-formula-assistant-arg div")[3].textContent).toBe(
          "f3 Arg2 def"
        );
      });
    });
  });

  describe("focus argument", () => {
    test("=FUNC1( focus index on 1st arg", async () => {
      await typeInComposer("=FUNC1(");
      expect(
        fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
          .textContent
      ).toBe("f1Arg1");
    });

    test("=FUNC1(42 focus index on 1st arg", async () => {
      await typeInComposer("=FUNC1(42");
      expect(
        fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
          .textContent
      ).toBe("f1Arg1");
    });

    test("=FUNC1(42 then add ',' focus index on 2nd arg", async () => {
      await typeInComposer("=FUNC1(42");
      await typeInComposer(",", false);
      expect(
        fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
          .textContent
      ).toBe("f1Arg2");
    });

    test("=FUNC1(42, focus index on 2nd arg", async () => {
      await typeInComposer("=FUNC1(42,");
      expect(
        fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
          .textContent
      ).toBe("f1Arg2");
    });

    test("functions with more arguments than allowed do not have focus", async () => {
      await typeInComposer("=FUNC1(42, 24, 22");
      expect(
        fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")
      ).toHaveLength(0);
    });

    describe("functions with repeatable argument always have a focus", () => {
      test("=FUNC3(84, focus on 2nd argument", async () => {
        await typeInComposer("=FUNC3(84,");
        expect(
          fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
            .textContent
        ).toBe("f3Arg2");
      });

      test("=FUNC3(84, 42, focus on 2nd argument", async () => {
        await typeInComposer("=FUNC3(84, 42,");
        expect(
          fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
            .textContent
        ).toBe("f3Arg2");
      });
    });

    describe("functions with more than one repeatable argument have an alternate focus", () => {
      test("=UPTOWNFUNC(1, 2, focus on 3th argument", async () => {
        await typeInComposer("=UPTOWNFUNC(1, 2,");
        expect(
          fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
            .textContent
        ).toBe("f4Arg3");
      });

      test("=UPTOWNFUNC(1, 2, 3, focus on 2nd argument", async () => {
        await typeInComposer("=UPTOWNFUNC(1, 2, 3,");
        expect(
          fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
            .textContent
        ).toBe("f4Arg2");
      });

      test("=UPTOWNFUNC(1, 2, 3, 4,  focus on 3th argument", async () => {
        await typeInComposer("=UPTOWNFUNC(1, 2, 3, 4,");
        expect(
          fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
            .textContent
        ).toBe("f4Arg3");
      });

      test("=UPTOWNFUNC(1, 2, 3, 4, 5, focus on 4th argument", async () => {
        await typeInComposer("=UPTOWNFUNC(1, 2, 3, 4, 5,");
        expect(
          fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
            .textContent
        ).toBe("f4Arg2");
      });

      test("Formula Helper updates when navigating with keyboard arrows", async () => {
        await typeInComposer("=UPTOWNFUNC(1, 2, 3");
        expect(
          fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
            .textContent
        ).toBe("f4Arg3");
        await keyDown({ key: "ArrowLeft" });
        await keyDown({ key: "ArrowLeft" });
        await keyDown({ key: "ArrowLeft" });
        await keyUp({ key: "ArrowLeft" });
        expect(
          fixture.querySelectorAll(".o-formula-assistant-arg.o-formula-assistant-focus span")[0]
            .textContent
        ).toBe("f4Arg2");
      });
    });
  });
});

describe("formula assistant for boolean functions", () => {
  beforeAll(() => {
    clearFunctions();
    functionRegistry.add("TRUE", {
      description: "TRUE",
      args: [],
      compute: () => true,
    });
    functionRegistry.add("FALSE", {
      description: "FALSE",
      args: [],
      compute: () => false,
    });
  });

  afterAll(() => {
    restoreDefaultFunctions();
  });

  test.each(["=TRUE(", "=true(", "=FALSE(", "=false("])(
    "show boolean formula assistant",
    async () => {
      await typeInComposer("=TRue(");
      expect(fixture.querySelectorAll(".o-formula-assistant")).toHaveLength(1);
    }
  );
});
