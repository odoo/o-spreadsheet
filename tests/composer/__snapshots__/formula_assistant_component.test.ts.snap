// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formula assistant appearance simple snapshot with =FUNC1( 1`] = `
<div
  class="o-formula-assistant-container user-select-none shadow pe-auto"
>
  <div
    class="o-formula-assistant"
  >
    <div
      class="o-formula-assistant-head"
    >
      <span>
        FUNC1
      </span>
       ( 
      
      <span
        class="o-formula-assistant-focus"
      >
        <span>
          
          <span>
            f1Arg1
          </span>
          
          
        </span>
      </span>
      <span>
        , 
      </span>
      <span>
        <span>
          
          <span>
            f1Arg2
          </span>
          
          
        </span>
      </span>
      
      
       ) 
    </div>
    <div
      class="o-formula-assistant-core pb-3 m-3"
    >
      <div
        class="o-formula-assistant-gray"
      >
        ABOUT
      </div>
      <div>
        func1 def
      </div>
    </div>
    <div
      class="o-formula-assistant-arg p-3 pt-0 display-flex flex-column o-formula-assistant-gray o-formula-assistant-focus"
    >
      <div>
        <span>
          f1Arg1
        </span>
        
        
        
      </div>
      <div
        class="o-formula-assistant-arg-description"
      >
        f1 Arg1 def
      </div>
    </div>
    <div
      class="o-formula-assistant-arg p-3 pt-0 display-flex flex-column o-formula-assistant-gray"
    >
      <div>
        <span>
          f1Arg2
        </span>
        
        
        
      </div>
      <div
        class="o-formula-assistant-arg-description"
      >
        f1 Arg2 def
      </div>
    </div>
    
    
  </div>
</div>
`;
